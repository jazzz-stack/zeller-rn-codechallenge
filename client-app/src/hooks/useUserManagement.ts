import {useState, useEffect, useCallback} from 'react';
import {useQuery} from '@apollo/client';
import {ZellerCustomer, ZellerCustomerInput, UserRole} from '../types/User';
import {LocalDatabaseService} from '../services/LocalDatabaseService';
import {listZellerCustomers} from '../graphql/queries';

export const useUserManagement = () => {
  const [customers, setCustomers] = useState<ZellerCustomer[]>([]);
  const [filteredCustomers, setFilteredCustomers] = useState<ZellerCustomer[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [searchQuery, setSearchQuery] = useState('');
  const [selectedFilter, setSelectedFilter] = useState<UserRole>('ALL');

  // GraphQL query to fetch customers from API
  const {loading: apiLoading, error: apiError, refetch} = useQuery(listZellerCustomers, {
    fetchPolicy: 'network-only',
    skip: true, // Don't run automatically, only when refetch is called
  });

  // Load customers from local database on mount first
  const loadLocalCustomers = useCallback(async () => {
    try {
      setLoading(true);
      const localCustomers = await LocalDatabaseService.getCustomers();
      setCustomers(localCustomers);
    } catch (err) {
      setError('Failed to load customers from local database');
      console.error('Error loading local customers:', err);
    } finally {
      setLoading(false);
    }
  }, []);

  // Sync customers from API to local database
  const syncFromAPI = useCallback(async () => {
    try {
      setLoading(true);
      setError(null);
      
      const response = await refetch();
      const apiCustomers = response.data?.listZellerCustomers?.items || [];
      console.log('API Customers:', apiCustomers);
      
      // Save to local database
      await LocalDatabaseService.saveCustomers(apiCustomers);
      
      // Update local state
      setCustomers(apiCustomers);
    } catch (err) {
      setError('Failed to sync customers from API');
      console.error('Error syncing from API:', err);
    } finally {
      setLoading(false);
    }
  }, [refetch]);

  // Add new customer
  const addCustomer = useCallback(async (customerInput: ZellerCustomerInput) => {
    try {
      setLoading(true);
      const newCustomer = await LocalDatabaseService.addCustomer(customerInput);
      setCustomers(prev => [...prev, newCustomer]);
    } catch (err) {
      setError('Failed to add customer');
      throw err;
    } finally {
      setLoading(false);
    }
  }, []);

  // Update existing customer
  const updateCustomer = useCallback(async (id: string, updates: Partial<ZellerCustomerInput>) => {
    try {
      setLoading(true);
      const updatedCustomer = await LocalDatabaseService.updateCustomer(id, updates);
      setCustomers(prev =>
        prev.map(customer =>
          customer.id === id ? updatedCustomer : customer
        )
      );
    } catch (err) {
      setError('Failed to update customer');
      console.error('Error updating customer:', err);
      throw err;
    } finally {
      setLoading(false);
    }
  }, []);

  // Delete customer
  const deleteCustomer = useCallback(async (id: string) => {
    try {
      setLoading(true);
      await LocalDatabaseService.deleteCustomer(id);
      setCustomers(prev => prev.filter(customer => customer.id !== id));
    } catch (err) {
      setError('Failed to delete customer');
      console.error('Error deleting customer:', err);
      throw err;
    } finally {
      setLoading(false);
    }
  }, []);

  // Apply filters and search
  useEffect(() => {
    let filtered = customers;

    // Apply role filter
    if (selectedFilter !== 'ALL') {
      filtered = filtered.filter(customer => customer.role === selectedFilter);
    }

    // Apply search filter
    if (searchQuery.trim()) {
      const query = searchQuery.toLowerCase();
      filtered = filtered.filter(
        customer =>
          customer.name.toLowerCase().includes(query) ||
          customer.email.toLowerCase().includes(query)
      );
    }

    setFilteredCustomers(filtered);
  }, [customers, selectedFilter, searchQuery]);

  // Load local customers on mount
  useEffect(() => {
    loadLocalCustomers();
  }, [loadLocalCustomers]);

  return {
    customers: filteredCustomers,
    loading: loading || apiLoading,
    error: error || (apiError?.message ?? null),
    searchQuery,
    selectedFilter,
    setSearchQuery,
    setSelectedFilter,
    addCustomer,
    updateCustomer,
    deleteCustomer,
    syncFromAPI,
    refreshLocalData: loadLocalCustomers,
  };
};
