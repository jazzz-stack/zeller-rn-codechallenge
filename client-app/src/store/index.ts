import {configureStore} from '@reduxjs/toolkit';
import customersReducer from './slices/customersSlice';

export const store = configureStore({
  reducer: {
    customers: customersReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: ['persist/PERSIST', 'persist/REHYDRATE'],
      },
    }),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
