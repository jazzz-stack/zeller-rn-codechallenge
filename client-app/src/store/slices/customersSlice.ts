import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {ZellerCustomer, ZellerCustomerInput, UserRole} from '../../types/User';
import {LocalDatabaseService} from '../../services/LocalDatabaseService';
import {apolloClient} from '../../config/apollo';
import {listZellerCustomers} from '../../graphql/queries';

interface CustomersState {
  customers: ZellerCustomer[];
  filteredCustomers: ZellerCustomer[];
  loading: boolean;
  error: string | null;
  searchQuery: string;
  selectedFilter: UserRole;
}

const initialState: CustomersState = {
  customers: [],
  filteredCustomers: [],
  loading: false,
  error: null,
  searchQuery: '',
  selectedFilter: 'ALL',
};

// Async thunks
export const loadCustomersFromLocal = createAsyncThunk(
  'customers/loadFromLocal',
  async (_, {rejectWithValue}) => {
    try {
      const customers = await LocalDatabaseService.getCustomers();
      return customers;
    } catch (error) {
      console.error('loadCustomersFromLocal error:', error);
      return rejectWithValue('Failed to load customers from local database');
    }
  },
);

export const syncCustomersFromAPI = createAsyncThunk(
  'customers/syncFromAPI',
  async (_, {rejectWithValue}) => {
    try {
      // Syncing customers from API...
      const response = await apolloClient.query({
        query: listZellerCustomers,
        fetchPolicy: 'network-only',
      });
      const apiCustomers = response.data?.listZellerCustomers?.items || [];
      // Customers saved to local database
      await LocalDatabaseService.saveCustomers(apiCustomers);
      return apiCustomers;
    } catch (error: any) {
      // Only log non-network errors to avoid console spam when offline
      const isNetworkError =
        error.networkError ||
        error.message?.includes('Network request failed') ||
        error.message?.includes('timeout') ||
        error.message?.includes('fetch') ||
        error.name === 'ApolloError';

      if (!isNetworkError) {
        console.error('syncCustomersFromAPI error:', error);
      }

      // Provide more specific error messages based on different error types
      if (
        error.networkError ||
        error.message?.includes('Network request failed')
      ) {
        return rejectWithValue(
          'No internet connection. Please check your network and try again.',
        );
      } else if (error.message?.includes('timeout')) {
        return rejectWithValue(
          'Network timeout. Please check your connection and try again.',
        );
      } else if (error.message?.includes('fetch')) {
        return rejectWithValue(
          'Connection failed. Please check your internet connection.',
        );
      } else if (error.name === 'ApolloError') {
        return rejectWithValue(
          'Network error occurred. Please check your internet connection.',
        );
      } else {
        return rejectWithValue('Failed to sync customers. Please try again.');
      }
    }
  },
);

// Background sync that doesn't affect UI error state
export const backgroundSyncCustomersFromAPI = createAsyncThunk(
  'customers/backgroundSyncFromAPI',
  async () => {
    try {
      // Background syncing customers from API...
      const response = await apolloClient.query({
        query: listZellerCustomers,
        fetchPolicy: 'network-only',
      });
      const apiCustomers = response.data?.listZellerCustomers?.items || [];
      await LocalDatabaseService.saveCustomers(apiCustomers);
      return apiCustomers;
    } catch (error: any) {
      // Suppress network error logs for background sync to avoid console spam
      const isNetworkError =
        error.networkError ||
        error.message?.includes('Network request failed') ||
        error.message?.includes('timeout') ||
        error.message?.includes('fetch') ||
        error.name === 'ApolloError';

      if (isNetworkError) {
        // Silently handle network errors in background sync
        console.log('Background sync skipped - no internet connection');
      } else {
        // Log non-network errors for debugging
        console.log('Background sync failed (non-network error):', error);
      }

      // Return empty array instead of rejecting to avoid UI errors
      return [];
    }
  },
);

export const addCustomer = createAsyncThunk(
  'customers/add',
  async (customerInput: ZellerCustomerInput, {rejectWithValue}) => {
    try {
      const newCustomer = await LocalDatabaseService.addCustomer(customerInput);
      return newCustomer;
    } catch (error) {
      return rejectWithValue('Failed to add customer');
    }
  },
);

export const updateCustomer = createAsyncThunk(
  'customers/update',
  async (
    {id, updates}: {id: string; updates: Partial<ZellerCustomerInput>},
    {rejectWithValue},
  ) => {
    try {
      const updatedCustomer = await LocalDatabaseService.updateCustomer(
        id,
        updates,
      );
      return updatedCustomer;
    } catch (error) {
      return rejectWithValue('Failed to update customer');
    }
  },
);

export const deleteCustomer = createAsyncThunk(
  'customers/delete',
  async (id: string, {rejectWithValue}) => {
    try {
      await LocalDatabaseService.deleteCustomer(id);
      return id;
    } catch (error) {
      return rejectWithValue('Failed to delete customer');
    }
  },
);

const customersSlice = createSlice({
  name: 'customers',
  initialState,
  reducers: {
    setSearchQuery: (state, action: PayloadAction<string>) => {
      state.searchQuery = action.payload;
      customersSlice.caseReducers.applyFilters(state);
    },
    setSelectedFilter: (state, action: PayloadAction<UserRole>) => {
      state.selectedFilter = action.payload;
      customersSlice.caseReducers.applyFilters(state);
    },
    clearError: state => {
      state.error = null;
    },
    applyFilters: state => {
      let filtered = [...state.customers];
      // Apply role filter
      if (state.selectedFilter !== 'ALL') {
        filtered = filtered.filter(customer => {
          return (
            customer.role.toUpperCase() === state.selectedFilter.toUpperCase()
          );
        });
      }

      // Apply search filter
      if (state.searchQuery.trim()) {
        const query = state.searchQuery.toLowerCase();
        filtered = filtered.filter(
          customer =>
            customer.name.toLowerCase().includes(query) ||
            customer.role.toLowerCase().includes(query),
        );
      }
      state.filteredCustomers = filtered;
    },
  },
  extraReducers: builder => {
    builder
      // Load customers from local
      .addCase(loadCustomersFromLocal.pending, state => {
        // console.log('loadCustomersFromLocal.pending');
        state.loading = true;
        state.error = null;
      })
      .addCase(loadCustomersFromLocal.fulfilled, (state, action) => {
        // console.log('loadCustomersFromLocal.fulfilled - payload:', action.payload.length, 'customers');
        state.loading = false;
        state.customers = action.payload;
        customersSlice.caseReducers.applyFilters(state);
      })
      .addCase(loadCustomersFromLocal.rejected, (state, action) => {
        console.log('loadCustomersFromLocal.rejected - error:', action.payload);
        state.loading = false;
        state.error = action.payload as string;
      })
      // Sync customers from API
      .addCase(syncCustomersFromAPI.pending, state => {
        console.log('syncCustomersFromAPI.pending');
        state.loading = true;
        state.error = null;
      })
      .addCase(syncCustomersFromAPI.fulfilled, (state, action) => {
        console.log(
          'syncCustomersFromAPI.fulfilled - payload:',
          action.payload.length,
          'customers',
        );
        state.loading = false;
        state.customers = action.payload;
        customersSlice.caseReducers.applyFilters(state);
      })
      .addCase(syncCustomersFromAPI.rejected, (state, action) => {
        console.log('syncCustomersFromAPI.rejected - error:', action.payload);
        state.loading = false;
        state.error = action.payload as string;
      })
      // Add customer
      .addCase(addCustomer.pending, state => {
        state.loading = true;
        state.error = null;
      })
      .addCase(addCustomer.fulfilled, (state, action) => {
        state.loading = false;
        state.customers.push(action.payload);
        customersSlice.caseReducers.applyFilters(state);
      })
      .addCase(addCustomer.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      })
      // Update customer
      .addCase(updateCustomer.pending, state => {
        state.loading = true;
        state.error = null;
      })
      .addCase(updateCustomer.fulfilled, (state, action) => {
        state.loading = false;
        const index = state.customers.findIndex(
          customer => customer.id === action.payload.id,
        );
        if (index !== -1) {
          state.customers[index] = action.payload;
        }
        customersSlice.caseReducers.applyFilters(state);
      })
      .addCase(updateCustomer.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      })
      // Delete customer
      .addCase(deleteCustomer.pending, state => {
        state.loading = true;
        state.error = null;
      })
      .addCase(deleteCustomer.fulfilled, (state, action) => {
        state.loading = false;
        state.customers = state.customers.filter(
          customer => customer.id !== action.payload,
        );
        customersSlice.caseReducers.applyFilters(state);
      })
      .addCase(deleteCustomer.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload as string;
      })
      // Background sync from API (doesn't affect error state)
      .addCase(backgroundSyncCustomersFromAPI.fulfilled, (state, action) => {
        // Only update data if we actually got customers from API
        if (action.payload && action.payload.length > 0) {
          state.customers = action.payload;
          customersSlice.caseReducers.applyFilters(state);
        }
      });
  },
});

export const {setSearchQuery, setSelectedFilter, clearError} =
  customersSlice.actions;
export default customersSlice.reducer;
