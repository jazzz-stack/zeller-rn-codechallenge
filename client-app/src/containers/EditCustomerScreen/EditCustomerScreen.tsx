import React, {useState} from 'react';
import {ZellerCustomerInput} from '../../types/User';
import {EditCustomerScreenProps} from '../../navigation/schema_types/navigation-schema';
import {useAppDispatch} from '../../store/hooks';
import {updateCustomer} from '../../store/slices/customersSlice';
import {CustomerForm, ConfirmationModal} from '../../components/Molecules';

const EditCustomerScreen = ({navigation, route}: EditCustomerScreenProps) => {
  const dispatch = useAppDispatch();
  const {customer} = route.params;
  const [modalVisible, setModalVisible] = useState(false);
  const [modalType, setModalType] = useState<'success' | 'error'>('success');
  const [modalMessage, setModalMessage] = useState('');

  const handleSubmit = async (customerData: ZellerCustomerInput) => {
    try {
      await dispatch(updateCustomer({id: customer.id, updates: customerData})).unwrap();
      setModalType('success');
      setModalMessage('Customer updated successfully');
      setModalVisible(true);
    } catch (err) {
      setModalType('error');
      setModalMessage('Failed to update customer');
      setModalVisible(true);
    }
  };

  const handleModalConfirm = () => {
    setModalVisible(false);
    if (modalType === 'success') {
      navigation.goBack();
    }
  };

  return (
    <>
      <CustomerForm
        title="Edit Customer"
        customer={customer}
        onSubmit={handleSubmit}
        onCancel={() => navigation.goBack()}
        submitButtonText="Update Customer"
      />
      
      <ConfirmationModal
        visible={modalVisible}
        type={modalType}
        message={modalMessage}
        onConfirm={handleModalConfirm}
        showCancel={false}
      />
    </>
  );
};

export default EditCustomerScreen;
