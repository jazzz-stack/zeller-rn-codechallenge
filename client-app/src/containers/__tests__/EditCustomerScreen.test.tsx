import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react-native';
import EditCustomerScreen from '../EditCustomerScreen/EditCustomerScreen';
import { useAppDispatch } from '../../store/hooks';
import { updateCustomer } from '../../store/slices/customersSlice';

// Mock dependencies
jest.mock('../../store/hooks', () => ({
  useAppDispatch: jest.fn(),
}));

jest.mock('../../store/slices/customersSlice', () => ({
  updateCustomer: jest.fn(),
}));

jest.mock('../../components/Molecules', () => ({
  CustomerForm: ({ title, customer, submitButtonText, cancelButtonText, onSubmit, onCancel }: any) => {
    const React = require('react');
    const { View, Text, TouchableOpacity } = require('react-native');
    
    return React.createElement(View, { testID: 'customer-form' },
      React.createElement(Text, { testID: 'form-title' }, title),
      React.createElement(Text, { testID: 'customer-name' }, customer?.name || ''),
      React.createElement(Text, { testID: 'customer-email' }, customer?.email || ''),
      React.createElement(Text, { testID: 'customer-role' }, customer?.role || ''),
      React.createElement(TouchableOpacity, { 
        testID: 'submit-button',
        onPress: () => onSubmit({
          name: 'Updated Customer',
          email: 'updated@example.com',
          role: 'Manager'
        })
      }, React.createElement(Text, {}, submitButtonText)),
      React.createElement(TouchableOpacity, { 
        testID: 'cancel-button',
        onPress: onCancel
      }, React.createElement(Text, {}, cancelButtonText))
    );
  },
  ConfirmationModal: ({ visible, onConfirm, onCancel }: any) => {
    const React = require('react');
    const { View, Text, TouchableOpacity } = require('react-native');
    
    if (!visible) return null;
    
    return React.createElement(View, { testID: 'confirmation-modal' },
      React.createElement(TouchableOpacity, { 
        testID: 'modal-confirm-button',
        onPress: onConfirm
      }, React.createElement(Text, {}, 'Confirm')),
      onCancel ? React.createElement(TouchableOpacity, { 
        testID: 'modal-cancel-button',
        onPress: onCancel
      }, React.createElement(Text, {}, 'Cancel')) : null
    );
  },
}));

describe('EditCustomerScreen', () => {
  const mockDispatch = jest.fn();
  const mockNavigation = {
    goBack: jest.fn(),
    navigate: jest.fn(),
  };

  const mockCustomer = {
    id: '1',
    name: 'John Doe',
    email: 'john@example.com',
    role: 'Admin' as const,
  };

  const mockProps = {
    navigation: mockNavigation as any,
    route: {
      params: {
        customer: mockCustomer,
      },
    } as any,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  it('renders EditCustomerScreen with correct props', () => {
    render(<EditCustomerScreen {...mockProps} />);
    
    expect(screen.getByTestId('customer-form')).toBeTruthy();
    expect(screen.getByTestId('form-title')).toBeTruthy();
    expect(screen.getByTestId('submit-button')).toBeTruthy();
    expect(screen.getByTestId('cancel-button')).toBeTruthy();
  });

  it('passes customer data to CustomerForm', () => {
    render(<EditCustomerScreen {...mockProps} />);
    
    expect(screen.getByTestId('customer-name')).toBeTruthy();
    expect(screen.getByTestId('customer-email')).toBeTruthy();
    expect(screen.getByTestId('customer-role')).toBeTruthy();
    expect(screen.getByText('John Doe')).toBeTruthy();
    expect(screen.getByText('john@example.com')).toBeTruthy();
    expect(screen.getByText('Admin')).toBeTruthy();
  });

  it('passes correct props to CustomerForm', () => {
    render(<EditCustomerScreen {...mockProps} />);
    
    expect(screen.getByText('Edit Customer')).toBeTruthy(); // title
    expect(screen.getByText('Update Customer')).toBeTruthy(); // submit button text
    expect(screen.getByTestId('submit-button')).toBeTruthy();
    expect(screen.getByTestId('cancel-button')).toBeTruthy();
  });

  it('handles cancel action correctly', () => {
    render(<EditCustomerScreen {...mockProps} />);
    
    const cancelButton = screen.getByTestId('cancel-button');
    fireEvent.press(cancelButton);
    
    expect(mockNavigation.goBack).toHaveBeenCalledTimes(1);
  });

  it('handles successful customer update', async () => {
    const mockUnwrap = jest.fn().mockResolvedValue({});
    mockDispatch.mockReturnValue({
      unwrap: mockUnwrap,
    });

    render(<EditCustomerScreen {...mockProps} />);
    
    const submitButton = screen.getByTestId('submit-button');
    
    await act(async () => {
      fireEvent.press(submitButton);
    });

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(updateCustomer({
        id: '1',
        updates: {
          name: 'Updated Customer',
          email: 'updated@example.com',
          role: 'Manager'
        }
      }));
    });

    await waitFor(() => {
      expect(mockUnwrap).toHaveBeenCalled();
    });

    await waitFor(() => {
      expect(screen.getByTestId('confirmation-modal')).toBeTruthy();
    });
  });

  it('handles customer update failure', async () => {
    const mockUnwrap = jest.fn().mockRejectedValue(new Error('Network error'));
    mockDispatch.mockReturnValue({
      unwrap: mockUnwrap,
    });

    render(<EditCustomerScreen {...mockProps} />);
    
    const submitButton = screen.getByTestId('submit-button');
    
    await act(async () => {
      fireEvent.press(submitButton);
    });

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(updateCustomer({
        id: '1',
        updates: {
          name: 'Updated Customer',
          email: 'updated@example.com',
          role: 'Manager'
        }
      }));
    });

    await waitFor(() => {
      expect(screen.getByTestId('confirmation-modal')).toBeTruthy();
    });
  });

  it('navigates back when success modal confirm is pressed', async () => {
    const mockUnwrap = jest.fn().mockResolvedValue({});
    mockDispatch.mockReturnValue({
      unwrap: mockUnwrap,
    });

    render(<EditCustomerScreen {...mockProps} />);
    
    const submitButton = screen.getByTestId('submit-button');
    
    await act(async () => {
      fireEvent.press(submitButton);
    });

    await waitFor(() => {
      expect(screen.getByTestId('confirmation-modal')).toBeTruthy();
    });

    const confirmButton = screen.getByTestId('modal-confirm-button');
    fireEvent.press(confirmButton);

    expect(mockNavigation.goBack).toHaveBeenCalledTimes(1);
  });

  it('dispatches updateCustomer action with correct customer id and updates', async () => {
    const mockUnwrap = jest.fn().mockResolvedValue({});
    mockDispatch.mockReturnValue({
      unwrap: mockUnwrap,
    });

    render(<EditCustomerScreen {...mockProps} />);
    
    const submitButton = screen.getByTestId('submit-button');
    
    await act(async () => {
      fireEvent.press(submitButton);
    });

    expect(mockDispatch).toHaveBeenCalledWith(updateCustomer({
      id: '1',
      updates: {
        name: 'Updated Customer',
        email: 'updated@example.com',
        role: 'Manager'
      }
    }));
  });

  it('does not navigate back when error modal confirm is pressed', async () => {
    const mockUnwrap = jest.fn().mockRejectedValue(new Error('Network error'));
    mockDispatch.mockReturnValue({
      unwrap: mockUnwrap,
    });

    render(<EditCustomerScreen {...mockProps} />);
    
    const submitButton = screen.getByTestId('submit-button');
    
    await act(async () => {
      fireEvent.press(submitButton);
    });

    await waitFor(() => {
      expect(screen.getByTestId('confirmation-modal')).toBeTruthy();
    });

    const confirmButton = screen.getByTestId('modal-confirm-button');
    fireEvent.press(confirmButton);

    expect(mockNavigation.goBack).not.toHaveBeenCalled();
  });

  it('renders with different customer data', () => {
    const differentCustomer = {
      id: '2',
      name: 'Jane Smith',
      email: 'jane@example.com',
      role: 'Manager' as const,
    };

    const propsWithDifferentCustomer = {
      ...mockProps,
      route: {
        params: {
          customer: differentCustomer,
        },
      } as any,
    };

    render(<EditCustomerScreen {...propsWithDifferentCustomer} />);
    
    expect(screen.getByText('Jane Smith')).toBeTruthy();
    expect(screen.getByText('jane@example.com')).toBeTruthy();
    expect(screen.getByText('Manager')).toBeTruthy();
  });

  // Snapshot Tests
  it('matches snapshot when rendering with customer', () => {
    const component = render(<EditCustomerScreen {...mockProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with different customer data', () => {
    const differentCustomer = {
      id: '2',
      name: 'Jane Smith',
      email: 'jane@example.com',
      role: 'Manager' as const,
    };

    const propsWithDifferentCustomer = {
      ...mockProps,
      route: {
        params: {
          customer: differentCustomer,
        },
      } as any,
    };

    const component = render(<EditCustomerScreen {...propsWithDifferentCustomer} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
