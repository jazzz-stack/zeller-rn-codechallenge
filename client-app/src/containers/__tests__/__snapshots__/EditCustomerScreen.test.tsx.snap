// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCustomerScreen matches snapshot when rendering with customer 1`] = `
<View
  testID="customer-form"
>
  <Text
    testID="form-title"
  >
    Edit Customer
  </Text>
  <Text
    testID="customer-name"
  >
    John Doe
  </Text>
  <Text
    testID="customer-email"
  >
    john@example.com
  </Text>
  <Text
    testID="customer-role"
  >
    Admin
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="submit-button"
  >
    <Text>
      Update Customer
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="cancel-button"
  >
    <Text />
  </View>
</View>
`;

exports[`EditCustomerScreen matches snapshot with different customer data 1`] = `
<View
  testID="customer-form"
>
  <Text
    testID="form-title"
  >
    Edit Customer
  </Text>
  <Text
    testID="customer-name"
  >
    Jane Smith
  </Text>
  <Text
    testID="customer-email"
  >
    jane@example.com
  </Text>
  <Text
    testID="customer-role"
  >
    Manager
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="submit-button"
  >
    <Text>
      Update Customer
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="cancel-button"
  >
    <Text />
  </View>
</View>
`;
